vminitd/Sources/vminitd/TerminalIO.swift

```
@@ -35,30 +36,49 @@ final class TerminalIO: ManagedProcess.IO & Sendable {
         stdio: HostStdio,
         log: Logger?
     ) throws {
-        let pair = try Terminal.create()
-        self.parent = pair.parent
-        self.child = pair.child
         self.stdio = stdio
         self.log = log
 
-        let ptyHandle = child.handle
-        let useHandles = stdio.stdin != nil || stdio.stdout != nil
-        // We currently set stdin to the controlling terminal always, so
-        // it must be a valid pty descriptor.
-        process.stdin = useHandles ? ptyHandle : nil
-
-        let stdoutHandle = useHandles ? ptyHandle : nil
-        process.stdout = stdoutHandle
-        process.stderr = stdoutHandle
+        process.stdin = nil
+        process.stdout = nil
+        process.stderr = nil
     }
 
     func resize(size: Terminal.Size) throws {
         if self.stdio.stdin != nil {
-            try parent.resize(size: size)
+            try parent?.resize(size: size)
+        }
+    }
+
+    func start() throws {}
+
+    func attach(pid: Int32, fd: Int32) throws {
+        #if os(Linux)
+        let containerFd = CZ_pidfd_open(pid, 0)
+        guard containerFd != -1 else {
+            throw POSIXError.fromErrno()
         }
+
+        let hostFd = CZ_pidfd_getfd(containerFd, fd, 0)
+        guard Foundation.close(Int32(containerFd)) == 0 else {
+            self.log?.error("failed to close pidfd: \(POSIXError.fromErrno())")
```

We should actually throw here, with how it is today it will silently succeed to the user

---

I haven't looked at how `.error()` works, does it only log?

---

Correct yep